### DISCUSSION TITLE
Showing my own data

### DISCUSSION THREAD

**Discussion by ebassi00** [eProsima/Fast-DDS-spy#109] [Q&A]
Hi everybody,
I read a previous discussion about FastDDSSpy not showing some user's custom data, and I discovered that, in order to do that, you have to declare DynamicTypes.
I already have a pretty big project using FastDDS where all my messages are generated from IDL to C++ code (meaning I didn't use the "-typeobject *.idl"). Now I'd like to use the Spy / Monitor to visualize the content of each message published by some writer. Is there a way to do so without generating DynamicTypes for each message (they are more than one and could be sent over multiple times changing values)? Thank you!

**Comment by juanlofer-eprosima** [comment]
Hi @ebassi00 ,

I am afraid you will need to regenerate all types in order to achieve your goal. First of all, I encourage you to update your code to use Fast-DDS 3.x, as the XTypes module has been totally refactored (supporting now many complex types that were not before) and sharing types is now enabled by default. Please have a look at [this tutorial](https://fast-dds.docs.eprosima.com/en/latest/notes/migration_guide.html) for some guidance on the migration process. Also note that after the v3 release Fast-DDS-Gen generates all auxiliar type object support files by default (no need to provide `-typeobject`).

An alternative that would not require to regenerate types would be to create a DynamicType via C++ for each of your types, and then publish DynamicData (instead of static). I discourage following this path as it is more complex and might not even be supported for one of your types (as I said, given the limitations of the XTypes module present in versions prior to v3).

**Comment by ebassi00** [comment]
Hi @juanlofer-eprosima , thank you for your reply,
I think i'm a bit lost then. I'm currently using FastDDSGen 3.3.0 and the generation process from IDL to C++ generates all the files it should generate (i.e. HelloWorld.* HelloWorldPubSubTypes.* HelloWorldCdrAux.*) . Said that, what I do is istantiate an object of type HelloWorld and then I publish it onto some topic. I guess this is not the right way to produce some DynamicType so that the Spy / Monitor can retrieve the data but at this point I don't really know how to do that, can you please clarify it to me?

**Reply by juanlofer-eprosima** [reply]
No worries it's indeed a bit confusing. When I mentioned the v3 version I was referring to the Fast-DDS one. Summing up, my recommendation is that you use Fast-DDS-spy latest release [v1.1.0](https://github.com/eProsima/Fast-DDS-spy/releases/tag/v1.1.0), which leverages Fast-DDS [v3.0.1](https://github.com/eProsima/Fast-DDS/releases/tag/v3.0.1) release, and this Fast-DDS version is expected to work with types generated with Fast-DDS-Gen [v4.0.1](https://github.com/eProsima/Fast-DDS-Gen/releases/tag/v4.0.1) release.

If you manage to adapt your code to work with these versions, you will be able to visualize in the spy data published statically (as you described it is your case) with no extra changes.

**Reply by ebassi00** [reply]
There comes another problem, I'm currently using latest fast-dds version from ConanCenter (2.14.3) so I'm not able to use the newer v3.0.1

**Reply by juanlofer-eprosima** [reply]
In that case you will have to use the previous Fast-DDS-spy release [v0.4.0](https://github.com/eProsima/Fast-DDS-spy/releases/tag/v0.4.0). With this version (and its 2.x Fast-DDS and 3.x Fast-DDS-Gen dependencies), types need to be generated with `-typeobject`. Please have a look a [this tutorial](https://dds-recorder.readthedocs.io/en/v0.4.0/rst/tutorials/dynamic_types.html) for more information, you may focus on the DDS Publisher section. But once again, be aware that only "simple" types such as the HelloWorld.idl from the tutorial will work with this Fast-DDS version (e.g. Complete.idl from the tutorial would not work).

**Reply by ebassi00** [reply]
Alright then..I need to visualize more complex types than HelloWorld...do you know when will v3.0.1 be published on ConanCenter?

**Reply by juanlofer-eprosima** [reply]
We are not the maintainers of packages at ConanCenter, hence cannot answer. However we provide ready-to-use Docker images that can be downloaded at [eProsima](https://eprosima.com).
